scripts/clusterfuzz/run.py

```diff
@@ -0,0 +1,163 @@
+# fuzzing large wasm files (to reduce the overhead we have of launching many
+# processes per file), which is less of an issue on ClusterFuzz.
+MAX_RANDOM_SIZE = 15 * 1024
+
+# The prefix for fuzz files.
+FUZZ_FILENAME_PREFIX = 'fuzz-'
+
+# The prefix for flags files.
+FLAGS_FILENAME_PREFIX = 'flags-'
+
+# The name of the fuzzer (appears after FUZZ_FILENAME_PREFIX /
+# FLAGS_FILENAME_PREFIX).
+FUZZER_NAME_PREFIX = 'binaryen-'
+
+# The root directory of the bundle this will be in, which is the directory of
+# this very file.
+ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
+
+# The path to the wasm-opt binary that we run to generate testcases.
+FUZZER_BINARY_PATH = os.path.join(ROOT_DIR, 'bin', 'wasm-opt')
+
+# The path to the fuzz_shell.js script that will execute the wasm in each
+# testcase.
+JS_SHELL_PATH = os.path.join(ROOT_DIR, 'scripts', 'fuzz_shell.js')
+
+# The arguments we provide to wasm-opt to generate wasm files.
+FUZZER_ARGS = [
+    # Generate a wasm from random data.
+    '--translate-to-fuzz',
+    # Run some random passes, to further shape the random wasm we emit.
+    '--fuzz-passes',
+    # Enable all features but disable ones not yet ready for fuzzing. This may
+    # be a smaller set than fuzz_opt.py, as that enables a few experimental
+    # flags, while here we just fuzz with d8's --wasm-staging.
+    '-all',
+    '--disable-shared-everything',
+    '--disable-fp16',
+]
+
+
+# Returns the file name for fuzz or flags files.
+def get_file_name(prefix, index):
+    return f'{prefix}{FUZZER_NAME_PREFIX}{index}.js'
+
+
+# Returns the contents of a .js fuzz file, given particular wasm contents that
+# we want to be executed.
+def get_js_file_contents(wasm_contents):
+    # Start with the standard JS shell.
+    with open(JS_SHELL_PATH) as file:
+        js = file.read()
+
+    # Prepend the wasm contents, so they are used (rather than the normal
+    # mechanism where the wasm file's name is provided in argv).
+    wasm_contents = ','.join([str(c) for c in wasm_contents])
+    js = f'var binary = new Uint8Array([{wasm_contents}]);\n\n' + js
+    return js
+
+
+def main(argv):
+    # Parse the options. See
+    # https://google.github.io/clusterfuzz/setting-up-fuzzing/blackbox-fuzzing/#uploading-a-fuzzer
+    output_dir = '.'
+    num = 100
+    expected_flags = ['input_dir=', 'output_dir=', 'no_of_files=']
+    optlist, _ = getopt.getopt(argv[1:], '', expected_flags)
+    for option, value in optlist:
+        if option == '--output_dir':
+            output_dir = value
+        elif option == '--no_of_files':
+            num = int(value)
+
+    for i in range(1, num + 1):
+        input_data_file_path = os.path.join(output_dir, '%d.input' % i)
+        wasm_file_path = os.path.join(output_dir, '%d.wasm' % i)
+
+        # wasm-opt may fail to run in rare cases (when the fuzzer emits code it
+        # detects as invalid). Just try again in such a case.
```

Comment:
Huh, I've never seen this before. Should we put the retry logic in wasm-opt instead?

Reply:
In general we would like to error on such things - they may be actual bugs in our fuzzer generator. When we run `fuzz_opt.py`, we've been noticing and fixing all such bugs we find. But on ClusterFuzz, that isn't the point of the testing (and we don't even have a way to report this back to us, the Binaryen devs).